File: tests/D2ItemToolkit.Net.Tests/Fakes.cs
using System.Collections.Generic;

namespace D2ItemToolkit.Tests
{
    internal sealed class FakeStringTable : IStringTable
    {
        public readonly Dictionary<int, string> Entries = new Dictionary<int, string>();

        public FakeStringTable Add(int index, string text)
        {
            Entries[index] = text;
            return this;
        }

        /// <summary>
        /// The punctuation the engine pulls from the .tbl rather than hardcoding.
        /// Without these every line comes out empty, which is itself worth a test.
        /// </summary>
        public FakeStringTable WithPunctuation()
        {
            return Add(DescStringIds.Space, " ")
                .Add(DescStringIds.Colon, " to")
                .Add(DescStringIds.Newline, "\n")
                .Add(DescStringIds.ListComma, "\n")
                .Add(DescStringIds.Percent, "%")
                .Add(DescStringIds.Plus, "+")
                .Add(DescStringIds.To, "to");
        }

        public string GetByIndex(int index)
        {
            string text;
            return Entries.TryGetValue(index, out text) ? text : null;
        }
    }

    internal sealed class FakeStatCostTable : IItemStatCostTable
    {
        public readonly Dictionary<int, StatDescriptor> Stats = new Dictionary<int, StatDescriptor>();
        public readonly List<int> Order = new List<int>();
        public readonly Dictionary<int, IReadOnlyList<int>> Groups = new Dictionary<int, IReadOnlyList<int>>();
        public readonly HashSet<int> Missing = new HashSet<int>();

        public int SkillIdShift { get; set; }

        /// <summary>0x4e4c76 returns 0 when a wOpBase reaches or exceeds this.</summary>
        public int RowCount { get; set; }

        public FakeStatCostTable()
        {
            SkillIdShift = 6;
            RowCount = 512;
        }

        public FakeStatCostTable Add(StatDescriptor descriptor)
        {
            Stats[descriptor.StatId] = descriptor;
            Order.Add(descriptor.StatId);
            return this;
        }

        public FakeStatCostTable AddGroup(int descGrp, params int[] statIds)
        {
            Groups[descGrp] = new List<int>(statIds);
            return this;
        }

        /// <summary>A stat id in the priority order with no backing row.</summary>
        public FakeStatCostTable AddMissing(int statId)
        {
            Missing.Add(statId);
            Order.Add(statId);
            return this;
        }

        public bool TryGetStat(int statId, out StatDescriptor descriptor)
        {
            if (Missing.Contains(statId))
            {
                descriptor = null;
                return false;
            }

            return Stats.TryGetValue(statId, out descriptor);
        }

        public IReadOnlyList<int> StatIdsByDescPriority
        {
            get { return Order; }
        }

        public IReadOnlyList<int> GetStatsInDescGroup(int descGrp)
        {
            IReadOnlyList<int> members;
            return Groups.TryGetValue(descGrp, out members) ? members : null;
        }
    }

    internal sealed class FakeStatValues : IStatValueSource
    {
        public readonly Dictionary<int, int> BaseStats = new Dictionary<int, int>();
        public readonly Dictionary<int, int> PlayerStats = new Dictionary<int, int>();

        public readonly HashSet<int> ItemTypes = new HashSet<int>();

        public int PlayerClass { get; set; }

        public FakeStatValues()
        {
            PlayerClass = -1;
        }

        public FakeStatValues AddItemType(int itemTypeId)
        {
            ItemTypes.Add(itemTypeId);
            return this;
        }

        public bool IsItemOfType(int itemTypeId)
        {
            return ItemTypes.Contains(itemTypeId);
        }

        public FakeStatValues AddBase(int statId, int value)
        {
            BaseStats[statId] = value;
            return this;
        }

        public FakeStatValues AddPlayer(int statId, int value)
        {
            PlayerStats[statId] = value;
            return this;
        }

        public int GetBaseStatValue(int statId, int layer)
        {
            int value;
            return BaseStats.TryGetValue(statId, out value) ? value : 0;
        }

        public readonly Dictionary<int, int> ItemStats = new Dictionary<int, int>();

        public FakeStatValues AddItemStat(int statId, int value)
        {
            ItemStats[statId] = value;
            return this;
        }

        public bool DescribedUnitIsItem { get; set; }

        public bool ItemTableAllowsDurability { get; set; }

        /// <summary>Distinct from GetItemStatValue(73): see the interface doc.</summary>
        public int TxtMaxDurability { get; set; }

        public int GetTxtMaxDurability()
        {
            return TxtMaxDurability;
        }

        public int GetItemStatValue(int statId)
        {
            int value;
            return ItemStats.TryGetValue(statId, out value) ? value : 0;
        }

        public int GetPlayerStatValue(int statId)
        {
            int value;
            return PlayerStats.TryGetValue(statId, out value) ? value : 0;
        }
    }

    internal sealed class FakeSkillTable : ISkillTable
    {
        public readonly Dictionary<int, string> Names = new Dictionary<int, string>();
        public readonly Dictionary<int, int> Classes = new Dictionary<int, int>();

        public int RowCount { get; set; }

        public FakeSkillTable()
        {
            RowCount = 400;
        }

        public FakeSkillTable Add(int skillId, string name, int classId = -1)
        {
            Names[skillId] = name;
            Classes[skillId] = classId;
            return this;
        }

        public bool SkillExists(int skillId)
        {
            return Names.ContainsKey(skillId);
        }

        public string GetSkillName(int skillId)
        {
            string name;
            return Names.TryGetValue(skillId, out name) ? name : null;
        }

        public int GetSkillClass(int skillId)
        {
            int classId;
            return Classes.TryGetValue(skillId, out classId) ? classId : -1;
        }
    }

    internal sealed class FakeClassTable : ICharacterClassTable
    {
        public readonly Dictionary<int, string> AllSkills = new Dictionary<int, string>();
        public readonly Dictionary<int, string> Tabs = new Dictionary<int, string>();
        public readonly Dictionary<int, string> ClassOnly = new Dictionary<int, string>();

        public FakeClassTable AddAllSkills(int classId, string text)
        {
            AllSkills[classId] = text;
            return this;
        }

        public FakeClassTable AddTab(int classId, int tabIndex, string text)
        {
            Tabs[classId * 4 + tabIndex] = text;
            return this;
        }

        public FakeClassTable AddClassOnly(int classId, string text)
        {
            ClassOnly[classId] = text;
            return this;
        }

        public string GetAllSkillsText(int classId)
        {
            string text;
            return AllSkills.TryGetValue(classId, out text) ? text : null;
        }

        public string GetSkillTabText(int classId, int tabIndex)
        {
            string text;
            return Tabs.TryGetValue(classId * 4 + tabIndex, out text) ? text : null;
        }

        public string GetClassOnlyText(int classId)
        {
            string text;
            return ClassOnly.TryGetValue(classId, out text) ? text : null;
        }

        /// <summary>A class counts as present once it has any charstats string.</summary>
        public bool ClassExists(int classId)
        {
            return AllSkills.ContainsKey(classId)
                   || ClassOnly.ContainsKey(classId)
                   || Tabs.ContainsKey(classId * 4)
                   || Tabs.ContainsKey(classId * 4 + 1)
                   || Tabs.ContainsKey(classId * 4 + 2);
        }
    }

    internal sealed class FakeMonsterTable : IMonsterTypeTable
    {
        public readonly Dictionary<int, string> Types = new Dictionary<int, string>();
        public readonly Dictionary<int, string> Monsters = new Dictionary<int, string>();

        public FakeMonsterTable AddType(int id, string name)
        {
            Types[id] = name;
            return this;
        }

        public FakeMonsterTable AddMonster(int id, string name)
        {
            Monsters[id] = name;
            return this;
        }

        public bool MonsterTypeExists(int monsterTypeId)
        {
            return Types.ContainsKey(monsterTypeId);
        }

        public bool MonsterExists(int monsterId)
        {
            return Monsters.ContainsKey(monsterId);
        }

        public string GetMonsterTypeName(int monsterTypeId)
        {
            string name;
            return Types.TryGetValue(monsterTypeId, out name) ? name : null;
        }

        public string GetMonsterName(int monsterId)
        {
            string name;
            return Monsters.TryGetValue(monsterId, out name) ? name : null;
        }
    }

    internal sealed class FakeGameTime : IGameTimeProvider
    {
        public bool HasTime = true;
        public int Degrees;

        public bool TryGetTimeAngle(out int degrees)
        {
            degrees = Degrees;
            return HasTime;
        }
    }

    /// <summary>Packs a by-time stat value the way the game stores it.</summary>
    internal static class ByTime
    {
        public static int Pack(int period, int low, int high)
        {
            return (period & 3)
                   | (((low + 256) & 0x3FF) << 2)
                   | (((high + 256) & 0x3FF) << 12);
        }
    }

    internal static class Build
    {
        public static StatDescriptor Stat(int statId, int descFunc, int strPos, int descVal = 1,
            int priority = 0, int strNeg = 0, int str2 = 0, int valShift = 0)
        {
            var descriptor = new StatDescriptor();
            descriptor.StatId = statId;
            descriptor.DescFunc = descFunc;
            descriptor.DescVal = descVal;
            descriptor.DescStrPos = strPos;
            descriptor.DescStrNeg = strNeg;
            descriptor.DescStr2 = str2;
            descriptor.DescPriority = priority;
            descriptor.ValShift = valShift;
            return descriptor;
        }

        /// <summary>The packed (layer, stat) key ReconstructView produces.</summary>
        public static KeyValuePair<int, int> Entry(int statId, int value, int layer = 0)
        {
            return new KeyValuePair<int, int>(ItemStatReader.PackStatKey(layer, statId), value);
        }
    }
}

File: tests/D2ItemToolkit.Ts.Tests/Fakes.ts
import { ItemStatReader } from '../../src/D2ItemToolkit.Ts/src/Stats/ItemStatReader.js';
import {
  DescStringIds,
  type ICharacterClassTable,
  type IGameTimeProvider,
  type IItemStatCostTable,
  type IMonsterTypeTable,
  type ISkillTable,
  type IStatValueSource,
  type IStringTable,
  type StatDescriptor,
} from '../../src/D2ItemToolkit.Ts/src/Types.js';

export class FakeStringTable implements IStringTable {
  readonly entries = new Map<number, string>();

  add(index: number, text: string): FakeStringTable {
    this.entries.set(index, text);
    return this;
  }

  /**
   * The punctuation the engine pulls from the .tbl rather than hardcoding.
   * Without these every line comes out empty, which is itself worth a test.
   */
  withPunctuation(): FakeStringTable {
    return this.add(DescStringIds.Space, ' ')
      .add(DescStringIds.Colon, ' to')
      .add(DescStringIds.Newline, '\n')
      .add(DescStringIds.ListComma, '\n')
      .add(DescStringIds.Percent, '%')
      .add(DescStringIds.Plus, '+')
      .add(DescStringIds.To, 'to');
  }

  getByIndex(index: number): string | null {
    return this.entries.get(index) ?? null;
  }
}

export class FakeStatCostTable implements IItemStatCostTable {
  readonly stats = new Map<number, StatDescriptor>();
  readonly order: number[] = [];
  readonly groups = new Map<number, readonly number[]>();
  readonly missing = new Set<number>();

  skillIdShift = 6;

  /** 0x4e4c76 returns 0 when a wOpBase reaches or exceeds this. */
  rowCount = 512;

  add(descriptor: StatDescriptor): FakeStatCostTable {
    this.stats.set(descriptor.statId, descriptor);
    this.order.push(descriptor.statId);
    return this;
  }

  addGroup(descGrp: number, ...statIds: number[]): FakeStatCostTable {
    this.groups.set(descGrp, statIds.slice());
    return this;
  }

  /** A stat id in the priority order with no backing row. */
  addMissing(statId: number): FakeStatCostTable {
    this.missing.add(statId);
    this.order.push(statId);
    return this;
  }

  tryGetStat(statId: number): StatDescriptor | null {
    if (this.missing.has(statId)) {
      return null;
    }

    return this.stats.get(statId) ?? null;
  }

  get statIdsByDescPriority(): readonly number[] {
    return this.order;
  }

  // The C# returns null for an unknown group; the engine's own guard tests null OR an empty
  // list, and Types.ts declares the return non-nullable, so an unknown group comes back empty.
  getStatsInDescGroup(descGrp: number): readonly number[] {
    return this.groups.get(descGrp) ?? [];
  }
}

export class FakeStatValues implements IStatValueSource {
  readonly baseStats = new Map<number, number>();
  readonly playerStats = new Map<number, number>();

  readonly itemTypes = new Set<number>();

  playerClass = -1;

  addItemType(itemTypeId: number): FakeStatValues {
    this.itemTypes.add(itemTypeId);
    return this;
  }

  isItemOfType(itemTypeId: number): boolean {
    return this.itemTypes.has(itemTypeId);
  }

  addBase(statId: number, value: number): FakeStatValues {
    this.baseStats.set(statId, value);
    return this;
  }

  addPlayer(statId: number, value: number): FakeStatValues {
    this.playerStats.set(statId, value);
    return this;
  }

  getBaseStatValue(statId: number, _layer: number): number {
    return this.baseStats.get(statId) ?? 0;
  }

  readonly itemStats = new Map<number, number>();

  addItemStat(statId: number, value: number): FakeStatValues {
    this.itemStats.set(statId, value);
    return this;
  }

  describedUnitIsItem = false;

  itemTableAllowsDurability = false;

  /** Distinct from GetItemStatValue(73): see the interface doc. */
  txtMaxDurability = 0;

  getTxtMaxDurability(): number {
    return this.txtMaxDurability;
  }

  getItemStatValue(statId: number): number {
    return this.itemStats.get(statId) ?? 0;
  }

  getPlayerStatValue(statId: number): number {
    return this.playerStats.get(statId) ?? 0;
  }
}

export class FakeSkillTable implements ISkillTable {
  readonly names = new Map<number, string>();
  readonly classes = new Map<number, number>();

  rowCount = 400;

  add(skillId: number, name: string, classId = -1): FakeSkillTable {
    this.names.set(skillId, name);
    this.classes.set(skillId, classId);
    return this;
  }

  skillExists(skillId: number): boolean {
    return this.names.has(skillId);
  }

  getSkillName(skillId: number): string | null {
    return this.names.get(skillId) ?? null;
  }

  getSkillClass(skillId: number): number {
    return this.classes.get(skillId) ?? -1;
  }
}

export class FakeClassTable implements ICharacterClassTable {
  readonly allSkills = new Map<number, string>();
  readonly tabs = new Map<number, string>();
  readonly classOnly = new Map<number, string>();

  addAllSkills(classId: number, text: string): FakeClassTable {
    this.allSkills.set(classId, text);
    return this;
  }

  addTab(classId: number, tabIndex: number, text: string): FakeClassTable {
    this.tabs.set(classId * 4 + tabIndex, text);
    return this;
  }

  addClassOnly(classId: number, text: string): FakeClassTable {
    this.classOnly.set(classId, text);
    return this;
  }

  getAllSkillsText(classId: number): string | null {
    return this.allSkills.get(classId) ?? null;
  }

  getSkillTabText(classId: number, tabIndex: number): string | null {
    return this.tabs.get(classId * 4 + tabIndex) ?? null;
  }

  getClassOnlyText(classId: number): string | null {
    return this.classOnly.get(classId) ?? null;
  }

  /** A class counts as present once it has any charstats string. */
  classExists(classId: number): boolean {
    return (
      this.allSkills.has(classId) ||
      this.classOnly.has(classId) ||
      this.tabs.has(classId * 4) ||
      this.tabs.has(classId * 4 + 1) ||
      this.tabs.has(classId * 4 + 2)
    );
  }
}

export class FakeMonsterTable implements IMonsterTypeTable {
  readonly types = new Map<number, string>();
  readonly monsters = new Map<number, string>();

  addType(id: number, name: string): FakeMonsterTable {
    this.types.set(id, name);
    return this;
  }

  addMonster(id: number, name: string): FakeMonsterTable {
    this.monsters.set(id, name);
    return this;
  }

  monsterTypeExists(monsterTypeId: number): boolean {
    return this.types.has(monsterTypeId);
  }

  monsterExists(monsterId: number): boolean {
    return this.monsters.has(monsterId);
  }

  getMonsterTypeName(monsterTypeId: number): string | null {
    return this.types.get(monsterTypeId) ?? null;
  }

  getMonsterName(monsterId: number): string | null {
    return this.monsters.get(monsterId) ?? null;
  }
}

export class FakeGameTime implements IGameTimeProvider {
  hasTime = true;
  degrees = 0;

  getTimeAngle(): number | null {
    return this.hasTime ? this.degrees : null;
  }
}

/** Packs a by-time stat value the way the game stores it. */
export const ByTime = {
  pack(period: number, low: number, high: number): number {
    return (period & 3) | (((low + 256) & 0x3ff) << 2) | (((high + 256) & 0x3ff) << 12);
  },
} as const;

export interface StatOptions {
  descVal?: number;
  priority?: number;
  strNeg?: number;
  str2?: number;
  valShift?: number;
}

export const Build = {
  /** The zeroed DTO C#'s `new StatDescriptor()` produces. */
  emptyStat(): StatDescriptor {
    return {
      statId: 0,
      descPriority: 0,
      descFunc: 0,
      descVal: 0,
      descStrPos: 0,
      descStrNeg: 0,
      descStr2: 0,
      descGrp: 0,
      descGrpFunc: 0,
      descGrpVal: 0,
      descGrpStrPos: 0,
      descGrpStrNeg: 0,
      descGrpStr2: 0,
      valShift: 0,
      op: 0,
      opParam: 0,
      opBase: 0,
    };
  },

  stat(
    statId: number,
    descFunc: number,
    strPos: number,
    options: StatOptions = {},
  ): StatDescriptor {
    const descriptor = Build.emptyStat();
    descriptor.statId = statId;
    descriptor.descFunc = descFunc;
    descriptor.descVal = options.descVal ?? 1;
    descriptor.descStrPos = strPos;
    descriptor.descStrNeg = options.strNeg ?? 0;
    descriptor.descStr2 = options.str2 ?? 0;
    descriptor.descPriority = options.priority ?? 0;
    descriptor.valShift = options.valShift ?? 0;
    return descriptor;
  },

  /** The packed (layer, stat) key ReconstructView produces. */
  entry(statId: number, value: number, layer = 0): readonly [number, number] {
    return [ItemStatReader.packStatKey(layer, statId), value];
  },
} as const;
